### PR TITLE
Instructor: Edit question: Check for empty string in Numscale/MSQ/Rank (Options/Recipients) number fields #8683

### LINKED ISSUE #8683: Instructor: Edit question: Check for empty string in Numscale/MSQ/Rank (Options/Recipients) number fields
Noticed while working on #8611 

When editing existing questions for rank options/recipient, MSQ, Numscale questions, if we leave a number field blank upon editing and delete the question, then instead of the question getting deleted, a error page in shown.

- **Environment**: 
`master` branch

**Steps to reproduce:**
- Login as instructor
- Edit a existing session
- Edit a existing question from the type mentioned above (or, add a new question, then edit it)
- Edit a number field (e.g. - in the `Numscale` question, edit the max/min/step value and leave it blank)
- `Delete` the question (Don't save the changes before deleting)

**Expected behaviour:** Following the example of question title and other fields, the question should get deleted.

**Actual behaviour:** An error page is shown.

**Additional info:** This happens because of `NumberFormatException`, for passing an empty string (`""`) in `Integer.parseInt(...)`
Now e.g in the `Numscale` question for minimum value we have:
https://github.com/TEAMMATES/teammates/blob/8a974ee703c46f4b3280552f46b3e767eb561eab/src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java#L50-L54

Here we check if the `minScaleString` is null or not, but that is not sufficient. If we are editing a question, then the value of the variable may not be null (As it is pointing towards something), but still can be an empty string, which is why it is passing the assert and passing an empty string into `Integer.parseInt(...)`, causing this error.

**Possible solution:** We can check if the String length is > 0 or not. For the above example,

```java
String minScaleString = 
        HttpRequestHelper.getValueFromParamMap(requestParameters, 
                                               Const.ParamsNames.FEEDBACK_QUESTION_NUMSCALE_MIN); 
Assumption.assertNotNull("Null minimum scale", minScaleString); 
if (minScaleString.length() > 0) {
    int minScale = Integer.parseInt(minScaleString);
}
```

### PR DESCRIPTION
**Pull request by bqnguyen94**
Fixes #8683 

**Outline of Solution**

Execute `deleteQuestion` without creating an `updatedQuestion`.

### CODE REVIEW

**Comment by Crphang** [comment]
I think the issue #8683 needs to be more specific about what it is trying to solve. It cannot solve the `edit` case properly because it is blocked by #8646, as no meaningful backend validation messages can be returned to the user. But this PR will help solve the scenario described for the `delete` case.

**Comment by sukanta-27** [comment]
@Crphang @bqnguyen94 Sorry for not being clearer in the issue. My focus was indeed on the `delete` option because, as @bqnguyen94 pointed out, checking if every field is valid or not, is not necessary for deletion. So, this issue can and I guess should focus on that aspect only. 

However, for the `save` option, it looks like we don't get any error while saving for `Numscale` question, as the backend validation is in place. But we do get an error for `MSQ` as well as Rank option/recipient type question. which I assumed will be solved in later issues as I guess it requires the backend validation for each type of question mentioned above?

**Comment by bqnguyen94** [comment]
@LuLechuan can help me take a look? :D

**Comment by LuLechuan** [comment]
Also, since this PR is to fix a bug. It would be good to have a test that fails initially, but after you make the changes, it passes

**Comment by sukanta-27** [comment]
@LuLechuan 

Checked. And no, it doesn't delete the question after trying to save. It does update the question.

**Comment by bqnguyen94** [comment]
Hmm seems like something among the recent merges to `master` makes `InstructorStudentRecordsPageUiTest` very prone to fail.

**Comment by LuLechuan** [comment]
LGTM

**Comment by damithc** [comment]
@xpdavid can review this one?

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 1]
```diff
@@ -68,12 +67,24 @@ protected ActionResult execute() throws EntityDoesNotExistException {
                                             .getInstructorFeedbackEditLink(courseId, feedbackSessionName));
     }
 
-    private void deleteQuestion(FeedbackQuestionAttributes updatedQuestion) {
-        logic.deleteFeedbackQuestion(updatedQuestion.getId());
+    private void deleteQuestion() {
+        String questionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_QUESTION_ID, questionId);
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Can use `getNonNullRequestParamValue` instead

Comment (bqnguyen94) [reply src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Hmm I see at least 10 places in this file can use this change. Should I change them all?

Comment (whipermr5) [reply src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Yup!!

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 2]
```diff
@@ -68,12 +62,20 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
 
-    private void deleteQuestion(FeedbackQuestionAttributes updatedQuestion) {
-        logic.deleteFeedbackQuestion(updatedQuestion.getId());
+    private void deleteQuestion() {
+        String questionId = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Can remove the 2 blank lines in between the last 3 parameters. Rationale: these 3 parameters are used together and form one logical block. Also, it'd be nice if the order was the same as where the parameters are used (i.e. `questionNumber` then `feedbackSessionName` then `courseId`).

Comment (bqnguyen94) [reply src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: 👍 done!

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 3]
```diff
@@ -68,12 +62,18 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
+        String questionNumber = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_NUMBER);
+
+        logic.deleteFeedbackQuestion(questionId);
         statusToUser.add(new StatusMessage(Const.StatusMessages.FEEDBACK_QUESTION_DELETED, StatusMessageColor.SUCCESS));
-        statusToAdmin = "Feedback Question " + updatedQuestion.questionNumber + " for session:<span class=\"bold\">("
-                        + updatedQuestion.feedbackSessionName + ")</span> for Course <span class=\"bold\">["
-                        + updatedQuestion.courseId + "]</span> deleted.<br>";
+        statusToAdmin = "Feedback Question " + questionNumber + " for session:<span class=\"bold\">("
+                        + feedbackSessionName + ")</span> for Course <span class=\"bold\">["
+                        + courseId + "]</span> deleted.<br>";
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: 8-space indentation

**Comment by bqnguyen94** [comment]
@whipermr5 any other suggestions? :)

**Pull request merged**